src/order/conditionally_complete_lattice.lean

```diff
@@ -692,3 +716,137 @@ noncomputable instance with_top.with_bot.complete_lattice {α : Type*}
+  Sup (coe '' t : set α) = (@Sup s _ t : α) :=
+by simp [dif_pos h]
+
+end has_Sup
+
+section has_Inf
+
+/-- `has_Inf` structure on a nonempty open subset `s` of a conditionally complete linear
+order. This definition is non-canonical (it uses `default s`); it should be used only as
+here, as an auxiliary instance in the construction of the `conditionally_complete_linear_order`
+structure. -/
+noncomputable def subset_has_Inf [inhabited s] : has_Inf s := {Inf := λ t,
+if ht : Inf (coe '' t : set α) ∈ s then ⟨Inf (coe '' t : set α), ht⟩ else default s}
+
+local attribute [instance] subset_has_Inf
+
+@[simp] lemma subset_Inf_def [inhabited s] :
+  @Inf s _ = λ t,
+  if ht : Inf (coe '' t : set α) ∈ s then ⟨Inf (coe '' t : set α), ht⟩ else default s :=
+rfl
+
+lemma subset_Inf_of_within [inhabited s] {t : set s} (h : Inf (coe '' t : set α) ∈ s) :
+  Inf (coe '' t : set α) = (@Inf s _ t : α) :=
+by simp [dif_pos h]
+
+end has_Inf
+
+local attribute [instance] subset_has_Sup
+local attribute [instance] subset_has_Inf
+
+/-- For a nonempty subset of a conditionally complete linear order to be a conditionally complete
+linear order, it suffices that it contain the `Sup` of all its nonempty bounded-above subsets, and
+the `Inf` of all its nonempty bounded-below subsets. -/
+noncomputable def subset_conditionally_complete_linear_order [inhabited s]
+  (h_Sup : ∀ {t : set s} (ht : t.nonempty) (h_bdd : bdd_above t), Sup (coe '' t : set α) ∈ s)
+  (h_Inf : ∀ {t : set s} (ht : t.nonempty) (h_bdd : bdd_below t), Inf (coe '' t : set α) ∈ s) :
+  conditionally_complete_linear_order s :=
+{ le_cSup := begin
+    rintros t c h_bdd hct,
+    -- The following would be a more natural way to finish, but gives a "deep recursion" error:
+    -- simpa [subset_Sup_of_within (h_Sup t)] using (strict_mono_coe s).monotone.le_cSup_image hct h_bdd,
+    have := (strict_mono_coe s).monotone.le_cSup_image hct h_bdd,
+    rwa subset_Sup_of_within s (h_Sup ⟨c, hct⟩ h_bdd) at this,
+  end,
+  cSup_le := begin
+    rintros t B ht hB,
+    have := (strict_mono_coe s).monotone.cSup_image_le ht hB,
+    rwa subset_Sup_of_within s (h_Sup ht ⟨B, hB⟩) at this,
+  end,
+  le_cInf := begin
+    intros t B ht hB,
+    have := (strict_mono_coe s).monotone.le_cInf_image ht hB,
+    rwa subset_Inf_of_within s (h_Inf ht ⟨B, hB⟩) at this,
+  end,
+  cInf_le := begin
+    rintros t c h_bdd hct,
+    have := (strict_mono_coe s).monotone.cInf_image_le hct h_bdd,
+    rwa subset_Inf_of_within s (h_Inf ⟨c, hct⟩ h_bdd) at this,
+  end,
+  ..subset_has_Sup s,
+  ..subset_has_Inf s,
+  ..distrib_lattice.to_lattice s,
+  ..classical.DLO s }
+
+section Ioo
+variables {a b : α}
+
+/-- The `Sup` function on a nonempty open subinterval `Ioo a b` of a conditionally complete linear
+order takes values within `Ioo a b`, for all nonempty bounded-above sets. -/
+lemma Ioo_Sup_within {a b : α} ⦃t : set (Ioo a b)⦄ (ht : t.nonempty) (h_bdd : bdd_above t) :
```

Comment:
Doesn't it work for any `ord_connected` set?

Reply:
Yes!